components/codec/src/byte.rs

```diff
@@ -348,11 +348,18 @@ impl MemComparableByteCodec {
             let src_ptr_end = src_ptr.add(src_len);
 
             loop {
-                let src_ptr_next = src_ptr.add(MEMCMP_GROUP_SIZE + 1);
-                if std::intrinsics::unlikely(src_ptr_next > src_ptr_end) {
+                // Make sure the source buffer is the expected length before
+                // constructing the next source pointer. Note that it is UB to
+                // even _create_ a pointer that points beyond its allocation.
+                let ptrdiff = src_ptr_end.offset_from(src_ptr);
+                let ptroob = ptrdiff < MEMCMP_GROUP_SIZE as isize + 1;
+
+                if std::intrinsics::unlikely(ptroob) {
                     return Err(ErrorInner::eof().into());
                 }
 
+                let src_ptr_next = src_ptr.add(MEMCMP_GROUP_SIZE + 1);
```

Comment:
I don't get it. How can it be an undefined behavior? I think accessing the data it points to is undefined, but get the pointer address is pure a math calculation.

Reply:
I don't understand the rules enough to say for sure, but here's what miri says about this code: https://gist.github.com/brson/26d2111ce4c1a555d0e9a817c37ff646

I take this to mean that even creating the pointer is UB.

I can't find language in https://doc.rust-lang.org/beta/reference/behavior-considered-undefined.html to support the idea that even doing pointer math to create out-of-bounds pointers is undefined. The strongest it says is that _dereferencing_ dangling pointers is undefined, which is obvious to me.

cc @oli-obk @RalfJung sorry for pinging you so much, but maybe you can explain why it's not ok to do pointer math that produces a bad pointer here.

Reply:
Yeah doing pointer arithmetic on out of bounds pointers is UB in LLVM, see for example this blog post: https://blog.regehr.org/archives/1518 - "it is UB to create a pointer that lies more than one element outside of an allocated object. It is UB merely to create such a pointer, it does not need to be dereferenced. Also, it is still UB even if the overflowed pointer happens to refer to some other allocated object.". This is about C, but applies equally to Rust

Reply:
Get it. So it's an undefined behavior because the calculation can produce an overflow value.

Reply:
> I can't find language in https://doc.rust-lang.org/beta/reference/behavior-considered-undefined.html to support the idea that even doing pointer math to create out-of-bounds pointers is undefined.

That list only enumerates UB caused by language primitives. When you call a method, the method docs mention the UB conditions for that method. In the case of [`add`](https://doc.rust-lang.org/nightly/std/primitive.pointer.html#method.add), it says:
> Both the starting and resulting pointer must be either in bounds or one byte past the end of the same allocated object. Note that in Rust, every (stack-allocated) variable is considered a separate allocated object.

Reply:
Generally, *all* raw pointer arithmetic functions have this requirement that the input and output must be within the same allocation -- with the exception of the `wrapping_*` methods.

Also note that `offset_from/to` require that [both pointers must be within the same allocation](https://doc.rust-lang.org/nightly/std/primitive.pointer.html#method.offset_from).

Reply:
If you are asking "why does `add` say this is UB", the short answer is that "because this is what [LLVM requires for `getelementptr inbounds`](https://llvm.org/docs/LangRef.html#getelementptr-instruction), and the long answer is that this helps the compiler analyze which pointers could potentially point where.

Reply:
Maybe we should report targetted (to the intrinsic) errors instead of generic errors when they are caused by the intrinsic's shim?

Reply:
The intrinsic is an unstable internal implementation detail though -- how it is called and which intrinsic is used for what can change any time. Seems a bit strange to surface that in error messages?

Reply:
Oh right :/ the calling function's docs are important, not the bare intrinsic. Yea not sure what we can do to improve the miri error then.

Reply:
Thanks for the help @RalfJung @oli-obk